drivers/sensor/voltage_divider/voltage.c

```diff
@@ -37,11 +36,18 @@ static int fetch(const struct device *dev, enum sensor_channel chan)
 		return -ENOTSUP;
 	}
 
+	ret = pm_device_runtime_get(dev);
```

Comment:
I fully agree. I was also thinking about that, but because I'm not aware of any board which actually requires an additional delay, I didn't add it. But in case you are aware of any actual board which requires this, I would be very happy to see this feature getting added.

I think adding an additional device tree property, as proposed by you, is a very good solution. I would suggest to define the delay in microseconds instead of ms. I think that is a more realistic time scale, see also the `off-on-delay-us` property of regulators:
https://github.com/zephyrproject-rtos/zephyr/blob/80b3f15a035ac0fdb17c056c7d07321c5dc5414a/dts/bindings/regulator/regulator.yaml#L273-L275

I would keep it simple and just do a simple `k_sleep()` inside `pm_action()` right after `gpio_pin_set_dt(&config->gpio_power, 1)` as done for regulators:
https://github.com/zephyrproject-rtos/zephyr/blob/80b3f15a035ac0fdb17c056c7d07321c5dc5414a/drivers/regulator/regulator_common.c#L133-L137

Reply:
Good points.

> I would keep it simple and just do a simple k_sleep() inside pm_action() right after gpio_pin_set_dt(&config->gpio_power, 1) as done for regulators

That would be simpler, but it might not work right with system power management if I understand it right. [Device PM documentation](https://docs.zephyrproject.org/latest/services/pm/device.html) has a note that:
> When using [System Power Management](https://docs.zephyrproject.org/latest/services/pm/system.html#pm-system), device transitions can be run from the idle thread. As functions in this context cannot block, transitions that intend to use blocking APIs must check whether they can do so with [k_can_yield()](https://docs.zephyrproject.org/latest/kernel/services/threads/index.html#c.k_can_yield).

I read that it is basically recommended that the PM action function would not block. busy looping is option, but doing that for around 1 ms, which I (unfortunately) have a use case (power_gpio controlling FET controlling voltage divider supply. For some reason it is just so slow) is maybe too much.

But I'll try to create another PR on this.

Reply:
@mkaranki  Thank you for preparing a separate PR. Let me know when it is ready. I'm really curious to follow the discussion on that topic.

I checked a couple of other device drivers and almost every single one is performing blocking operations inside `pm_action()` callback. For many drivers it is just mandatory to do that. Basically every driver controlling any kind of peripheral chip attached over any bus like I2C or SPI will need to perform I2C/SPI transactions in order to suspend/resume the chip. Such transactions are by definition blocking and of course it is required to perform them inside the `pm_action()` callback, because you need to make sure that the chip is suspended/resumed before returning from the callback.

In this special case of the voltage divider you could actually defer the `sleep()` until `fetch()`, as you suggest, but conceptually that is wrong, because the resume function is, in my opinion, supposed to ensure that the resume is completed before returning. Deferring the `sleep()` is basically breaking the promise of the API that the device is resumed on return from the callback.

You are right that the documentation actually indicates that the callback is not allowed to block. I have no idea how that fits with all the existing drivers which are performing blocking operations. Maybe System Power Management is just not supported in combination with such drivers, as indicated by the doc: 

> This method of device power management can be useful in the following scenarios:
> Systems with no device requiring any blocking operations when suspending and resuming.

Maybe Device Runtime Power Management needs to be used in that case. As that is the preferred method anyway, maybe that is just the way to go.

I'm looking forward to your PR.

Reply:
There's https://github.com/zephyrproject-rtos/zephyr/pull/71193 on top of the branch on this PR at now. It's done without deferring the delay to fetch function for now. Small optimization is to initialize the device as suspended when the pm device runtime is enabled to avoid annoying delay on init.

While doing it, I noticed there's almost identical current_amp driver: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/current_amp/current_amp.c. The voltage-divider and current-amp drivers are almost identical of each other. Pity that the don't already share the common parts although the drivers are quite simple. There's maybe just not natural place for the shared code in the driver three. Anyways, I think the PM device runtime and power gpio changes would be good to add into that driver as well once they are ready.